src/digraphs.c

```diff
@@ -854,6 +854,85 @@ static Obj FuncDIGRAPH_TOPO_SORT(Obj self, Obj adj) {
   return out;
 }
 
+// TODO(*) use generic DFS, when we have one.
+
+// Assumes that <adj> is the list of adjacencies of a symmetric digraph
+static Obj FuncDIGRAPH_SYMMETRIC_SPANNING_FOREST(Obj self, Obj adj) {
+  UInt  nr, i, j, k, next, len, level;
+  Obj   nbs, out, out_j, new;
+  UInt  *stack, *ptr;
+
+  nr = LEN_PLIST(adj);
+
+  if (nr == 0) {
+    return NEW_PLIST(T_PLIST_EMPTY + IMMUTABLE, 0);
+  }
+
+  // init the adjacencies of the spanning forest
+  out = NEW_PLIST(T_PLIST_TAB + IMMUTABLE, nr);
+  SET_LEN_PLIST(out, nr);
+  for (i = 1; i <= nr; i++) {
+    SET_ELM_PLIST(out, i, NEW_PLIST(T_PLIST_EMPTY + IMMUTABLE, 0));
+    SET_LEN_PLIST(ELM_PLIST(out, i), 0);
+    CHANGED_BAG(out);
+  }
+
+  // init the buffer
+  ptr   = calloc(nr + 1, sizeof(UInt));
+  stack = malloc((2 * nr + 2) * sizeof(UInt));
+
+  for (i = 1; i <= nr; i++) {
+    if (!ptr[i] && LEN_LIST(ELM_PLIST(adj, i)) > 0) {
+      level    = 1;
+      stack[0] = i;
+      stack[1] = 1;
+
+      while (1) {
+        j = stack[0];
+        k = stack[1];
+        nbs = ELM_PLIST(adj, j);
+
+        if (ptr[j] || k > (UInt) LEN_LIST(nbs)) {
+          ptr[j] = 1; // vertex <j> and its descendents are now fully explored
+          if (--level == 0) {
+            break; // we've explored the whole graph from vertex <i>
+          }
+          stack -= 2; // otherwise backtrack
+          ptr[stack[0]] = 0;
+          stack[1]++;
+        } else {
+          ptr[j] = 1;
+          next = INT_INTOBJ(ADDR_OBJ(nbs)[k]);
+          level++;
+          stack += 2;
+          stack[0] = next;
+          stack[1] = 1;
+          // if <next> is a brand new vertex, add it to the spanning tree
+          if (ptr[next] == 0) {
+            // Create the forward edge, j -> next
+            out_j = ELM_PLIST(out, j);
+            len   = LEN_PLIST(out_j);
+            if (len == 0) {
+              RetypeBag(out_j, T_PLIST_CYC + IMMUTABLE);
+              CHANGED_BAG(out_j);
+            }
+            AssPlist(out_j, len + 1, INTOBJ_INT(next));
+            // Create the reverse edge, next -> j
+            // Since <next> is new, new = out[next] is still a T_PLIST_EMPTY
+            new = ELM_PLIST(out, next);
+            RetypeBag(new, T_PLIST_CYC + IMMUTABLE);
+            CHANGED_BAG(new);
+            AssPlist(new, 1, INTOBJ_INT(j));
```

Comment:
Going by my previous comment, the `CHANGED_BAG(new)` should be *replaced* by the `CHANGED_BAG(out)`. In this case, is it still necessary to have the `CHANGED_BAG` within the `if` statement a few lines above, since they're so close together?